service/history/workflow/update/update.go

```diff
@@ -281,9 +316,19 @@ func (u *Update) Admit(
 		return invalidArgf("unable to marshal request: %v", err)
 	}
 	u.request = reqAny
-	u.setState(stateProvisionallyAdmitted)
-	eventStore.OnAfterCommit(func(context.Context) { u.setState(stateAdmitted) })
-	eventStore.OnAfterRollback(func(context.Context) { u.setState(stateCreated) })
+	prevState := u.setState(stateProvisionallyAdmitted)
+	eventStore.OnAfterCommit(func(context.Context) {
+		if u.state != stateProvisionallyAdmitted {
+			return
+		}
+		u.setState(stateAdmitted)
+	})
+	eventStore.OnAfterRollback(func(context.Context) {
+		if u.state != stateProvisionallyAdmitted {
+			return
+		}
+		u.setState(prevState)
+	})
 	return nil
```

Comment:
These changes (and the similar ones below) look plausible, but can you explain why you have made them? It looks like you are fixing a bug. What would a test look like that demonstrates the bug you are fixing?